### PR TITLE
Update Path from PathSegments 

### LINKED ISSUE #3952: ArcSegment is not updated
Hello! I have following XAML:
```
<Path Margin="0,3,0,0">
    <Path.Data>
        <PathGeometry>
            <PathGeometry.Figures>
                <PathFigures>
                    <PathFigure IsClosed="False">
                        <PathFigure.Segments>
                            <PathSegments>
                                <ArcSegment SweepDirection="CounterClockwise" />
                            </PathSegments>
                        </PathFigure.Segments>
                    </PathFigure>
                </PathFigures>
            </PathGeometry.Figures>
        </PathGeometry>
    </Path.Data>
</Path>
```

I'm trying to update arc segment dynamically using binding to vm/code behind files. But arc segment keeps using initial parameters that I set on app start. Am I missing something?

### PR DESCRIPTION
**Pull request by jmacato**
## What does the pull request do?

Invalidates the parent `Path` control when a `PathSegment`-derived type changes value through bindings.

## What is the current behavior?

Currently we dont listen on any changes from the child `PathSegments`'s property changes, hence we dont handle the invalidation there.

## What is the updated/expected behavior with this PR?

Changing values to `PathSegment`-derived types should trigger an geometry invalidation.
 
## Fixed issues
Fixes #3952

### CODE REVIEW

File: tests/Avalonia.RenderTests/Shapes/PathTests.cs [hunk 1]
```diff
@@ -353,7 +353,47 @@ public async Task Path_With_PenLineCap()
             await RenderToFile(target);
             CompareImages();
         }
+        
+        [Fact]
+        public async Task PathSegment_Triggers_Invalidation_On_Property_Change()
```
Comment (jmacato) [inline tests/Avalonia.RenderTests/Shapes/PathTests.cs:358]: Need a better place for this test, any ideas @AvaloniaUI/core ?

Comment (donandren) [inline tests/Avalonia.RenderTests/Shapes/PathTests.cs:358]: i think a better place for the test is `GeometryTests `(or dedicated `PathGeometryTests`) in `Avalonia.Visuals.UnitTests` as this is not related to rendering

File: tests/Avalonia.RenderTests/Shapes/PathTests.cs [hunk 2]
```diff
@@ -353,7 +353,47 @@ public async Task Path_With_PenLineCap()
...
+                Background = Brushes.White,
+                Child =targetPath
+            };
+            
+            Assert.Equal(10, targetPath.Bounds.Height);
+            Assert.Equal(10, targetPath.Bounds.Width);
+
+            targetSegment.Size = new Size(20, 20);
 
+            Assert.Equal(20, targetPath.Bounds.Height);
```
Comment (donandren) [inline tests/Avalonia.RenderTests/Shapes/PathTests.cs:393]: test can be simplified something like:
``` C#
        [Fact]
        public void PathSegment_Triggers_Invalidation_On_Property_Change()
        {
            var targetSegment = new ArcSegment()
            {
                Size = new Size(10, 10),
                Point = new Point(5, 5)
            };

            var target = new PathGeometry
            {
                Figures = new PathFigures
                {
                    new PathFigure { IsClosed = false, Segments = new PathSegments { targetSegment } }
                }
            };
            bool changed = false;

            target.Changed += (s, e) => changed = true;

            targetSegment.Size = new Size(20, 20);

            Assert.True(changed);
        }
```

File: src/Avalonia.Visuals/Media/PathFigure.cs [hunk 3]
```diff
@@ -25,14 +28,41 @@ public sealed class PathFigure : AvaloniaObject
...
+
+        private IDisposable? _segmentsPropertiesObserver;
+        
         /// <summary>
         /// Initializes a new instance of the <see cref="PathFigure"/> class.
         /// </summary>
         public PathFigure()
         {
+            SegmentsProperty.Changed.AddClassHandler<PathFigure>((s, e) =>
+                s.OnSegmentsChanged(e.NewValue as PathSegments));
```
Comment (donandren) [inline src/Avalonia.Visuals/Media/PathFigure.cs:43]: need to go to static ctor, as  there will be leaks:
```C#
 static PathFigure()
        {
            SegmentsProperty.Changed.AddClassHandler<PathFigure>((s, e) =>
                s.OnSegmentsChanged(e.NewValue as PathSegments));
```

**Review by donandren — CHANGES_REQUESTED** [review]
Looking very good, but need some small tweaks

**Comment by jmacato** [comment]
@donandren please do another review pass :)

**Review by donandren — APPROVED** [review]
Looking great, tests on CI are passing also 👍

File: src/Avalonia.Visuals/Media/PathFigure.cs [hunk 4]
```diff
@@ -25,6 +28,12 @@ public sealed class PathFigure : AvaloniaObject
         public static readonly StyledProperty<Point> StartPointProperty
                         = AvaloniaProperty.Register<PathFigure, Point>(nameof(StartPoint));
 
+        internal event EventHandler SegmentsInvalidated;
+
+        private IDisposable? _segmentsObserver;
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/PathFigure.cs:33]: The `?` here and on the next line is producing a warning because nullable reference types aren't enabled for this file. Either add `#nullable enable` or remove the `?`.

File: src/Avalonia.Visuals/Media/PathFigure.cs [hunk 5]
```diff
@@ -33,6 +42,30 @@ public PathFigure()
             Segments = new PathSegments();
         }
 
+        static PathFigure()
+        {
+            SegmentsProperty.Changed.AddClassHandler<PathFigure>((s, e) =>
+                s.OnSegmentsChanged(e.NewValue as PathSegments));
+        }
+
+        private void OnSegmentsChanged(PathSegments? arg2NewValue)
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/PathFigure.cs:51]: Nit: put private methods at the bottom.

File: src/Avalonia.Visuals/Media/PathFigure.cs [hunk 6]
```diff
@@ -33,6 +42,30 @@ public PathFigure()
...
+        {
+            _segmentsObserver?.Dispose();
+            _segmentsPropertiesObserver?.Dispose();
+
+            _segmentsObserver = _segments?.ForEachItem(
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/PathFigure.cs:56]: Nit: this isn't an observable. I'd call it `_segmentsSubscription` or something.

File: src/Avalonia.Visuals/Media/PathFigure.cs [hunk 7]
```diff
@@ -33,6 +42,30 @@ public PathFigure()
...
+                _ => InvalidateSegments(),
+                _ => InvalidateSegments(),
+                InvalidateSegments);
+            
+            _segmentsPropertiesObserver = _segments?.TrackItemPropertyChanged(_ => InvalidateSegments());
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/PathFigure.cs:61]: Again, not an observable.

File: src/Avalonia.Visuals/Media/PathGeometry.cs [hunk 8]
```diff
@@ -104,12 +104,26 @@ private void OnFiguresChanged(PathFigures figures)
...
+                    s.SegmentsInvalidated -= InvalidateGeometryFromSegments;
+                    InvalidateGeometry();
+                },
+                InvalidateGeometry);
+            
             _figuresPropertiesObserver = figures?.TrackItemPropertyChanged(_ => InvalidateGeometry());
+ 
         }
 
+        public void InvalidateGeometryFromSegments(object _, EventArgs __)
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/PathGeometry.cs:123]: This can be private I think?

**Review by grokys — CHANGES_REQUESTED** [review]
Looking good, a couple of things that need fixing and a few nits.

File: src/Avalonia.Visuals/Media/PathFigure.cs [hunk 9]
```diff
@@ -25,6 +28,12 @@ public sealed class PathFigure : AvaloniaObject
         public static readonly StyledProperty<Point> StartPointProperty
                         = AvaloniaProperty.Register<PathFigure, Point>(nameof(StartPoint));
 
+        internal event EventHandler SegmentsInvalidated;
+
```
Comment (danwalmsley) [inline src/Avalonia.Visuals/Media/PathFigure.cs:32]: _segments needs to be up here too!

File: src/Avalonia.Visuals/Media/PathFigure.cs [hunk 10]
```diff
@@ -1,3 +1,6 @@
+using System;
+using System.ComponentModel;
+using Avalonia.Collections;
```
Comment (danwalmsley) [inline src/Avalonia.Visuals/Media/PathFigure.cs:3]: nullable enable here!

**Review by danwalmsley — CHANGES_REQUESTED** [review]

File: src/Avalonia.Visuals/Media/PathFigure.cs [hunk 11]
```diff
@@ -94,9 +133,7 @@ internal void ApplyTo(StreamGeometryContext ctx)
             ctx.EndFigure(IsClosed);
         }
 
-        private PathSegments _segments;
-
         public override string ToString()
```
Comment (danwalmsley) [inline src/Avalonia.Visuals/Media/PathFigure.cs:136]: public method above private methods

**Review by danwalmsley — CHANGES_REQUESTED** [review]
nit

File: src/Avalonia.Visuals/Media/PathFigure.cs (hunk 11 above)
Comment (danwalmsley) [reply src/Avalonia.Visuals/Media/PathFigure.cs]: fixed

**Review by danwalmsley — APPROVED** [review]

File: src/Avalonia.Visuals/Media/PathFigure.cs [hunk 12]
```diff
@@ -33,6 +51,31 @@ public PathFigure()
             Segments = new PathSegments();
```
Comment (grokys) [inline src/Avalonia.Visuals/Media/PathFigure.cs:51]: @jmacato I know this was a while ago, but do you remember why you marked `Segments` as nullable when it's always initialized with a non-null value in the constructor?

Comment (jmacato) [reply src/Avalonia.Visuals/Media/PathFigure.cs:51]: i think it was to suppress a null warning, but yeah it doesnt make sense if it's already initialized

Comment (grokys) [reply src/Avalonia.Visuals/Media/PathFigure.cs:51]: Actually looking at WPF, it seems that even though `Segments` is initially non-null, it can be set to null so I think it's correct.
